File: mod/Nova.Core/Communications/GroundStation.cs
using System;
using System.Collections.Generic;
using Nova.Core.Components.Communications;
using Nova.Core.Utils;

namespace Nova.Core.Communications;

// Hard-coded ground-station endpoints. Today this is just KSC; if
// more stations land, they get cfg-driven via the same code path
// rather than expanding the hard-coded set.
public static class GroundStation {

  // KSC on Kerbin. Latitude/longitude/altitude per stock KSP launch
  // pad coordinates; body parameters per the homeworld definition.
  private const double KscLatitudeDeg = -0.0972;
  private const double KscLongitudeDeg = -74.5577;
  private const double KscAltitudeM = 75;
  private const double KerbinRadiusM = 600_000;
  private const double KerbinSiderealPeriodS = 21549.425;

  // KSC's deep-space-network-class transmitter. Tx and Gain set
  // several orders of magnitude above any vessel-mounted antenna so
  // the homeworld stays the dominant link partner; RefDistance set
  // to outer-system reach so distant probes still close a usable
  // link to KSC at full hardware rate.
  private const double KscTxPower = 1.0e9;
  private const double KscGain = 1.0e3;
  private const double KscMaxRate = 1.0e9;
  private const double KscRefDistance = 1.0e10;

  // Build the KSC endpoint. `homeworldPositionAt` supplies the
  // body's centre at a given UT — KSP wraps Planetarium; tests can
  // pass a constant. Keeps this file Nova.Core-pure.
  public static Endpoint Ksc(Func<double, Vec3d> homeworldPositionAt) {
    return SurfaceEndpoint(
      id: "KSC",
      latitudeDeg: KscLatitudeDeg,
      longitudeDeg: KscLongitudeDeg,
      altitudeM: KscAltitudeM,
      bodyRadiusM: KerbinRadiusM,
      siderealPeriodS: KerbinSiderealPeriodS,
      bodyPositionAt: homeworldPositionAt,
      antennas: new List<Antenna> {
        new() {
          TxPower = KscTxPower,
          Gain = KscGain,
          MaxRate = KscMaxRate,
          RefDistance = KscRefDistance,
        },
      });
  }

  // Generic surface-mounted endpoint. The body is assumed to spin
  // around its Y axis (KSP convention: polar axis = +Y); rotation
  // is eastward (positive longitude direction with positive UT).
  // Body-fixed (lat, lon, alt) → body-frame Cartesian → world frame
  // via bodyPositionAt(ut) translation.
  public static Endpoint SurfaceEndpoint(
      string id,
      double latitudeDeg, double longitudeDeg, double altitudeM,
      double bodyRadiusM, double siderealPeriodS,
      Func<double, Vec3d> bodyPositionAt,
      List<Antenna> antennas) {

    var latRad = latitudeDeg * Math.PI / 180.0;
    var lonRad = longitudeDeg * Math.PI / 180.0;
    var radius = bodyRadiusM + altitudeM;
    var omega = 2 * Math.PI / siderealPeriodS;
    var cosLat = Math.Cos(latRad);
    var sinLat = Math.Sin(latRad);

    Vec3d Position(double ut) {
      var phase = lonRad + omega * ut;
      var local = new Vec3d(
        radius * cosLat * Math.Cos(phase),
        radius * sinLat,
        radius * cosLat * Math.Sin(phase));
      return bodyPositionAt(ut) + local;
    }

    var ep = new Endpoint { Id = id, PositionAt = Position };
    ep.Antennas.AddRange(antennas);
    return ep;
  }
}

File: crates/nova-sim/src/comms/ground_station.rs
//! Ground-station endpoint factories. Mirrors
//! `mod/Nova.Core/Communications/GroundStation.cs`.
//!
//! The C# version takes a `bodyPositionAt: Func<UT, Vec3d>` closure
//! because endpoints owned a closure-shaped position evaluator. The
//! Rust port is data-driven — surface endpoints store
//! `EndpointKind::Ground { primary, lat, lon, alt }` and look up the
//! body's centre through `World::ephemeris` on each `position_at`
//! call. Constructors here produce a `GroundStationSpec` (raw data)
//! that `CommsSystem::add_ground_station` materialises into an
//! `Endpoint` once a fresh `Ground(n)` id has been minted.

use crate::ephem::BodyId;

use super::antenna::Antenna;
use super::endpoint::{Endpoint, EndpointId, EndpointKind, PathSummary};
use super::motion::MotionModel;

/// Static spec for a surface-mounted endpoint. Becomes an `Endpoint`
/// once `CommsSystem` mints a `Ground(n)` id for it.
#[derive(Clone, Debug)]
pub struct GroundStationSpec {
    pub name: String,
    pub primary: BodyId,
    pub latitude_deg: f64,
    pub longitude_deg: f64,
    pub altitude_m: f64,
    pub antennas: Vec<Antenna>,
}

impl GroundStationSpec {
    /// Materialise as a fully-populated `Endpoint` with the given
    /// ground-counter `id`. `motion` carries a `Surface` hint;
    /// `primary_body` is set to the parent body so the occluder set
    /// resolves correctly.
    pub fn into_endpoint(self, id: u32) -> Endpoint {
        let GroundStationSpec {
            name,
            primary,
            latitude_deg,
            longitude_deg,
            altitude_m,
            antennas,
        } = self;
        Endpoint {
            id: EndpointId::Ground(id),
            name,
            kind: EndpointKind::Ground { primary, latitude_deg, longitude_deg, altitude_m },
            motion: Some(MotionModel::Surface {
                parent: primary,
                latitude_deg,
                longitude_deg,
                altitude_m,
            }),
            primary_body: Some(primary),
            antennas,
            is_predictable: true,
            path_to_home: PathSummary::default(),
        }
    }
}

/// KSC on `kerbin`. Lat/lon/alt match stock-KSP launch-pad coordinates.
/// Tx/Gain/MaxRate/RefDistance set several orders of magnitude above
/// any vessel-mounted antenna so the homeworld stays the dominant link
/// partner; RefDistance reaches outer-system so distant probes still
/// close a usable link to KSC at full hardware rate.
pub fn ksc(kerbin: BodyId) -> GroundStationSpec {
    GroundStationSpec {
        name: "KSC".to_string(),
        primary: kerbin,
        latitude_deg: -0.0972,
        longitude_deg: -74.5577,
        altitude_m: 75.0,
        antennas: vec![Antenna {
            tx_power: 1.0e9,
            gain: 1.0e3,
            max_rate: 1.0e9,
            ref_distance: 1.0e10,
        }],
    }
}
